### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by Qoosim**
This PR contains of all the logics to play the tic-tac-toe game.
It also contains method to checkmate any invalid move attempt by either of the participants

### CODE REVIEW

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello team  🌟

**Very nice work on this project, however, the following issues need to be fixed before moving on ⬇️** 

1. Please separate each class into a different file and put them inside the `lib` folder, ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)), ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).

2. The `main.rb` should use the already defined classes in the `lib` folder to start the game and it should contain all the `puts` and `gets.chomp`, ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).

3. The game can determine the winner successfully, but you have to make it stop the gameplay and displays a message that a player has won as soon as he has a combination, right now even if I win it would let me still play and switch turns until we finish all the empty spots and then it will say that I have won, fix that please.

4. Use `attr_readers/attr_writers` instead of `attr_accessors`. By using `attr_accessors` you allow writing that property from outside the class and that variable can be changed to anything, ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)).

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

File: bin/main [hunk 1]
```diff
@@ -1,14 +1,16 @@
 #!/usr/bin/env ruby
 #rubocop:disable all
```
Comment (ebukaume) [inline bin/main]: Remove!

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  def display_board(board)
+    puts " #{board[0]} | #{board[1]} | #{board[2]} "
+    puts '-----------'
+    puts " #{board[3]} | #{board[4]} | #{board[5]} "
+    puts '-----------'
+    puts " #{board[6]} | #{board[7]} | #{board[8]} "
+    puts ' '
+  end
+
```
Comment (ebukaume) [inline lib/board.rb]: `puts` only allowed in `main.rb` file as stated above.

File: lib/game_status.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+#rubocop:disable all
```
Comment (ebukaume) [inline lib/game_status.rb]: Remove!

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+#rubocop:disable all
```
Comment (ebukaume) [inline lib/board.rb]: Remove!

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+#rubocop:disable all
```
Comment (ebukaume) [inline lib/player.rb]: Remove!

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    end
+
+    if game_status.won?(game_board.board, game_status.win_combinations)
+      puts "Congratulations #{game_status.winner(game_board.board, tic_tac_game.name1, tic_tac_game.name2, game_status.win_combinations)} , You Won"
+    end
+
+    if game_status.draw?(game_board.board, game_status.win_combinations)
+      puts "#{tic_tac_game.name1},#{tic_tac_game.name2} Game has been a Draw Well Played "
+    end
+  end
```
Comment (ebukaume) [inline lib/player.rb]: Return the message and use `puts` inside `main.rb` to display the message.

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    puts 'enter no to take position 1-9'
+    input = gets.strip
+    index = game_board.input_to_index(input)
+    if game_board.valid_move?(game_board.board, index)
+      game_board.move(game_board.board, index, player)
+      game_board.display_board(game_board.board)
+    else
+      puts 'invalid move Type again'
+     end
+  end
```
Comment (ebukaume) [inline lib/player.rb]: Same as above...

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes Required! :stop_sign: 

Hi team :wave: 

Nice implementing the changes requested by my colleague. However, there are some issues we need to fix.

- [ ] You should not have `puts/print` and `get.chomp` outside main.rb file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] **Why did you configure a linter if you are going to ignore them?** Please remove all the `#rubocop:disable all` directives in your code so that your code would be properly linted!

_Kindly see below for more..._

**Please submit another code review request after attending to those issues.**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

File: .rubocop.yml [hunk 8]
```diff
@@ -1,31 +1,50 @@
...
 AllCops:
   Exclude:
     - "bin/*"
     - "db/**/*"
     - "config/**/*"
     - "Guardfile"
     - "Rakefile"
-    - "node_modules/**/*"
   DisplayCopNames: true
+  DisabledByDefault: true
```
Comment (imhta) [inline .rubocop.yml]: You have excluded everything and disabled the cops
It is should not be like this.

**Review by imhta — CHANGES_REQUESTED** [review]
# Changes required 🛑 

### Hello  @TashfeenRao @Qoosim  , 👋 You have done a great job in this project 👏 

* [ ] Why did you configure a linter if you are going to ignore them? In `.stickler.yml` don't exclude your project directory. Fix that
----
### Don't forgot to request for another code review
----
**Pon Muthu Selvam N**
__nponmuthuselvam@gmail.com__
_If you have any queries, feel free to ping me on slack_

File: .rubocop.yml [hunk 9]
```diff
@@ -1,31 +1,51 @@
...
 
 Metrics/LineLength:
-  Max: 120
+  Enabled: false
+  Max: 300
 Metrics/MethodLength:
+  Enabled: false
```
Comment (ebukaume) [inline .rubocop.yml]: You are not supposed to modify anything in this file

File: bin/main [hunk 10]
```diff
@@ -1,175 +1,61 @@
...
+    puts 'enter no to take position 1-9'
+    input = gets.strip
+    index = game_board.input_to_index(input)
+    if game_board.valid_move?(game_board.board, index)
+      game_board.move(game_board.board, index, player)
+      tic_tac_game.display_board(game_board.board)
+    else
+      puts 'invalid move Type again'
+     end
   end
```
Comment (ebukaume) [inline bin/main:52]: `line 42` is always going to be flagged by rubocop. You just have to learn to write cleaner code. Please let's do the right thing.

File: lib/player.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    @player = 'X'
+  end
+
+  def Play(game_board, tic_tac_game, game_status, game_player)
+    tic_tac_game.turn(game_board, game_player, tic_tac_game.name1, tic_tac_game.name2, game_status.win_combinations, tic_tac_game) until game_status.game_over?(game_board.board, game_status.win_combinations)
+
+    return "Congratulations #{game_status.winner(game_board.board, tic_tac_game.name1, tic_tac_game.name2, game_status.win_combinations)} , You Won" if game_status.won?(game_board.board, game_status.win_combinations)
+
+    return "#{tic_tac_game.name1},#{tic_tac_game.name2} Game has been a Draw Well Played " if game_status.draw?(game_board.board, game_status.win_combinations)
+  end
```
Comment (ebukaume) [inline lib/player.rb]: Learn to keep your character length to a max of between 80 and 90. It will help you in the long run.

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes Required! :stop_sign: 

Hi team :wave: 

I really do not understand why you guys keep going back and forth on this. Please use the right configuration provided [here](https://github.com/microverseinc/linters-config/tree/master/ruby) by microverse. 

**Kindly submit another code review request affect makig that fix**

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

**Review by Redvanisation — CHANGES_REQUESTED** [review]
# Changes required  ♻️ 

### Hello team  🌟

**Very nice work on this project, however, the following issues need to be fixed before moving on ⬇️** 

_ Copy [this](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) file and paste it into your `rubocop.yml` file exactly like it is **without any modifications**, you will then have about **5** linting errors, please fix then if you don't know how then you could also use `rubocop -a` command it would help you a bit with automatic fixes, but you need to fix anything else that doesn't get auto-fixed.

### _After making the required changes please submit a new code review request_

### Good luck! 🍀

**Review by Redvanisation — APPROVED** [review]
# Project approved ✔️ 


### _Well done guys! it seems that you have met the requirements so I am approving your project. 👍_ 

### You may merge to master and submit a project completion form. 🚀 

### Good luck! 🍀

**Pull request merged**